### ISSUE TITLE
Some data problems in nuplan_v0.2 and nuplan_v0.2_mini

### ISSUE THREAD

**Issue by ChenYingbing** [motional/nuplan-devkit#14]
Hi, nuPlan Team,

I guess I found some problems when reading the data from both nuplan_v0.2 and nuplan_v0.2_mini.

It is about the module of **AgentsFeatureBuilder()**, I used the matplolib to show the data of **features["agents"]**. And for position (x, y) data in **features["agents"].ego**, anything goes well, the x-y curves are clear.

However, when I deal with the x, y data in **features["agents"].agents**, I found that almost all position data among frames keep unchanged, even when the vx or vy data is not zero.

Part of extracted data are as following, where the position of x, y are keep constant, while the vx and vy are [-2.5884e+00,  2.6541e-01]

```
tensor([[ 2.2393e+01,  4.5293e+01,  9.4689e-01, -2.5884e+00,  2.6541e-01,
         -1.1345e-15,  4.0554e-01,  4.1339e-01],
        [ 2.2393e+01,  4.5293e+01,  9.4689e-01, -2.5884e+00,  2.6541e-01,
         -2.2732e-16,  4.0554e-01,  4.1339e-01],
        [ 2.2393e+01,  4.5293e+01,  9.4689e-01, -2.5884e+00,  2.6541e-01,
         -2.2732e-16,  4.0554e-01,  4.1339e-01],
        [ 2.2393e+01,  4.5293e+01,  9.4689e-01, -2.5884e+00,  2.6541e-01,
         -2.2732e-16,  4.0554e-01,  4.1339e-01],
        [ 2.2393e+01,  4.5293e+01,  9.4689e-01, -2.5884e+00,  2.6541e-01,
          2.4507e-15,  4.0554e-01,  4.1339e-01]])
```

Thank you for your time and work. I am not sure whether it is due to my fault in code or this is really a bug in dataset. If it is my fault, it would be grateful if you can point out where goes wrong. Thanks again, my test code is as following:

```
import os

import torch
import torch.nn as nn
import torchvision
import torchvision.transforms as transforms

# scenario_builder
from nuplan.planning.scenario_builder.nuplan_db.nuplan_scenario_builder import NuPlanScenarioBuilder
from nuplan.planning.scenario_builder.scenario_filter import ScenarioFilters

from nuplan.planning.simulation.trajectory.trajectory_sampling import TrajectorySampling
from nuplan.planning.training.data_loader.datamodule import DataModule
from nuplan.planning.training.data_loader.log_splitter import LogSplitter

# fitures
from nuplan.planning.training.preprocessing.feature_caching_preprocessor import FeatureCachingPreprocessor
from nuplan.planning.training.preprocessing.features.raster import Raster
from nuplan.planning.training.preprocessing.features.trajectory import Trajectory
from nuplan.planning.training.preprocessing.features.vector_map import VectorMap
from nuplan.planning.training.preprocessing.features.agents_trajectories import AgentsTrajectories
from nuplan.planning.training.preprocessing.features.agents import AgentsFeature

# feature_builders
from nuplan.planning.training.preprocessing.feature_builders.raster_feature_builder import RasterFeatureBuilder
from nuplan.planning.training.preprocessing.feature_builders.vector_map_feature_builder import VectorMapFeatureBuilder
from nuplan.planning.training.preprocessing.feature_builders.agents_feature_builder import AgentsFeatureBuilder

# target builders
from nuplan.planning.training.preprocessing.target_builders.ego_trajectory_target_builder import \
    EgoTrajectoryTargetBuilder

from nuplan.planning.training.preprocessing.test.dummy_vectormap_builder import DummyVectorMapBuilder
from nuplan.planning.utils.multithreading.worker_pool import WorkerPool
from nuplan.planning.utils.multithreading.worker_ray import RayDistributed
from nuplan.planning.utils.multithreading.worker_sequential import Sequential

# DATA_VERSION = "nuplan_v0.2"
DATA_VERSION = "nuplan_v0.2_mini"

class DataLoader():
    """
    Tests data loading functionality
    """

    def load_data(self, scenario_builder: NuPlanScenarioBuilder, worker: WorkerPool) -> None:
        """
        Tests that the training dataloader can be iterated without errors
        """
        # Create splitter
        splitter = LogSplitter(
            log_splits={
                'train': ["2021.05.26.20.05.14_38_1622073985538950.8_1622074969538793.5"],
                'val': ["2021.06.04.19.10.47_47_1622848319071793.5_1622849413071686.2"],
                'test': ["2021.05.28.21.56.29_24_1622239057169313.0_1622240664170207.2"],
            })

        # Create feature builder
        # @note can add feature by adding a XXXFeatureBuilder
        map_radius = 20.0 # meters
        features = FeatureCachingPreprocessor(cache_dir=None,
                                              force_feature_computation=True,
                                              # ficture builders
                                              feature_builders=[
                                                  DummyVectorMapBuilder(),
                                                  VectorMapFeatureBuilder(radius=map_radius),
                                                  RasterFeatureBuilder(
                                                      map_features={
                                                          'LANE': 255, 'INTERSECTION': 255,
                                                          'STOP_LINE': 128, 'CROSSWALK': 128},
                                                      num_input_channels=4,
                                                      target_width=224,
                                                      target_height=224,
                                                      target_pixel_size=0.5,
                                                      ego_width=2.297,
                                                      ego_front_length=4.049,
                                                      ego_rear_length=1.127,
                                                      ego_longitudinal_offset=0.0,
                                                      baseline_path_thickness=1,
                                                  ),
                                                  AgentsFeatureBuilder(
                                                    TrajectorySampling(
                                                      num_poses=4, 
                                                      time_horizon=1.5) 
                                                  )],
                                              # target builders
                                              target_builders=[
                                                  EgoTrajectoryTargetBuilder(TrajectorySampling(
                                                      num_poses=12,
                                                      time_horizon=6.0)
                                                  )]
                                              )

        # # Extract scenarios: example from all_scnearios.yaml
        # log_names: null                   # List of logs from which scenarios should be included
        # log_labels: null                  # Labels used to filter logs
        # max_scenarios_per_log: null       # Maximum number of scenarios extracted per log
        # scenario_types: null              # A list of scenario types to be used
        # scenario_tokens: null             # A list of scenario tokens to be used
        # map_name: null                    # Filter out scenarios on a specific map
        # # Filters
        # shuffle: false                    # 1. Shuffle selected scenarios
        # limit_scenarios_per_type: null    # 2. How many scenarios per type to keep (float = fraction, int = num)
        # subsample_ratio: null             # 3. Subsample a scenario relative to the database frequency (e.g. nominal freq @ 20Hz means 400 examples for 20s -> 2.5% subsample_ratio means 10 examples for 20s)
        # flatten_scenarios: false          # 4. Flatten scenarios (e.g. 10x 20-samples scenarios to 200x 1-sample scenarios)
        # remove_invalid_goals: false       # 5. Remove scenarios with invalid goals (e.g on an intersection, exceeded max dist)
        # limit_total_scenarios: null       # 6. How many total scenarios to keep (float = fraction, int = num)

        scenario_filter = ScenarioFilters(
            log_names=None,
            log_labels=None,
            max_scenarios_per_log=None,
            scenario_types=None,
            scenario_tokens=None,
            map_name=None,
            shuffle=False,
            limit_scenarios_per_type=None, # 100
            subsample_ratio=0.05,
            flatten_scenarios=True,
            remove_invalid_goals=True,
            limit_total_scenarios=1000,
        )

        scenarios = scenario_builder.get_scenarios(scenario_filter, worker)
        assert len(scenarios) > 0, "error1"

        # Construct data module
        batch_size = 4
        num_workers = 4
        datamodule = DataModule(
            feature_and_targets_builders=features,
            splitter=splitter,
            train_fraction=1.0,
            val_fraction=0.1,
            test_fraction=0.1,
            all_scenarios=scenarios,
            dataloader_params={"batch_size": batch_size, "num_workers": num_workers, "drop_last": True},
        )

        # Initialize data module
        datamodule.setup('fit')

        assert len(datamodule.train_dataloader()) > 0, "error2"

        # Run
        print("Start with scenario_num=", len(scenarios), 
              "; len_train=", len(datamodule.train_dataloader()))
        cii = 0
        for features, targets in datamodule.train_dataloader():
            # Validate that all features and targets are preset

            # Validate the dimensions
            scenario_features: Raster = features["raster"]
            ego_target_traj: Trajectory = targets["trajectory"]
            agent_target_trajs: AgentsFeature = features["agents"]
            vector_map : VectorMap = features["vector_map"]

            if True:
              print("============================")
              print("data[%d]: batch_siz=%d." %(cii, batch_size)) # batch_siz = 4

              print(">>> features")
              print(features.keys())
              """
              coords: List[<np.ndarray: num_lane_segments, 2, 2>].
                  The (x, y) coordinates of the start and end point of the lane segments.

                  coords: List[FeatureDataType]
              """
              print("vector_map len(coord)", len(vector_map.coords)) # = 4 (batch siz)
              if len(vector_map.coords) > 0:
                print("coord0:", vector_map.coords[0].size()) # torch.Size([7031, 2, 2])

              """
              multi_scale_connections: List[Dict of {scale: connections_of_scale}].
                  Each connections_of_scale is represented by an array of <np.ndarray: num_connections, 2>,
                  and each column in the array is [from_lane_segment_idx, to_lane_segment_idx].
              
                  multi_scale_connections: List[Dict[int, FeatureDataType]]
              """
              print("vector_map len(connections)", len(vector_map.multi_scale_connections)) # = 4 (batch siz)
              if len(vector_map.multi_scale_connections) > 0:
                print("connection0:", vector_map.multi_scale_connections[0].keys()) # dict_keys([1])
                print("connection0[key = 1]:", vector_map.multi_scale_connections[0][1].size()) # ([7008, 2])
                # print("details= ", vector_map.multi_scale_connections[0][1]) # [[from_lane_segment_idx, to_lane_segment_idx], ...]

              print("ratster", scenario_features.data.size()) # ratster torch.Size([batch_siz, 4, 224, 224])
              #
              # print("ego_layer shape=", scenario_features.ego_layer.size()) # torch.Size([batch_siz, 224, 224])
              # print("agents_layer shape=", scenario_features.agents_layer.size()) # torch.Size([batch_siz, 224, 224])
              # print("roadmap_layer shape=", scenario_features.roadmap_layer.size()) # torch.Size([batch_siz, 224, 224])
              # print("baseline_paths_layer shape=", scenario_features.baseline_paths_layer.size()) # torch.Size([batch_siz, 224, 224])

              # agent_target_trajs.ego = [batch_siz, ]
              # self.assertEqual(len(feature.ego), self.batch_size)
              # self.assertEqual(len(feature.ego[0]), self.num_total_past_poses)
              # self.assertEqual(len(feature.ego[0][0]), AgentsFeature.ego_state_dim())
              # self.assertEqual(len(feature.agents), self.batch_size)
              # self.assertEqual(len(feature.agents[0]), self.num_total_past_poses)
              # self.assertEqual(len(feature.agents[0][0]), self.num_agents)
              # self.assertEqual(len(feature.agents[0][0][0]), AgentsFeature.agents_states_dim())
              print("agent_trajs[ego]", len(agent_target_trajs.ego), # =4 batch size
                                        len(agent_target_trajs.ego[0]), # =21(20+1) num of poses
                                        len(agent_target_trajs.ego[0][0]) # =3 [x, y, heading]
              )
              print("agent_trajs[others]", len(agent_target_trajs.agents), # =4 batch size
                                           len(agent_target_trajs.agents[0]), # =21(20+1) num of poses
                                           len(agent_target_trajs.agents[0][0]), # =129/136/134/128 num of other agents
                                           len(agent_target_trajs.agents[0][0][0]) # =8 [x, y, heading, ....]
              )

              # ego: List[<np.ndarray: num_frames, 3>].
              #   The outer list is the batch dimension.
              #   The num_frames includes both present and past frames.
              #   The last dimension is the ego pose (x, y, heading) at time t.
              #   Example dimensions: 8 (batch_size) x 5 (1 present + 4 past frames) x 3
              print("agent_trajs[ego]", len(agent_target_trajs.ego), # =4 batch size
                                        agent_target_trajs.ego[0].size() # torch.Size([21, 3]), [num_poses, 3 (x, y, heading)]
              )
              # agents: List[<np.ndarray: num_frames, num_agents, 8>].
              #     The outer list is the batch dimension.
              #     The num_frames includes both present and past frames.
              #     The num_agents is padded to fit the largest number of agents across all frames. !!!
              #     The last dimension is the agent pose (x, y, heading) velocities (vx, vy, yaw rate) 
              #         and size (length, width) at time t.
              print("agent_trajs[agents]", len(agent_target_trajs.agents), # =4 batch size
                                           agent_target_trajs.agents[0].size() # torch.Size([21, Variable, 8]), 
                                                                               #    [num_poses, max_num_of_agents, 8]
              )

              # # agent_target_trajs: AgentsFeature = features["agents"]
              for ai in range(0, agent_target_trajs.num_agents_in_sample(0)):
                # xs = agent_target_trajs.agents[0][:, ai, 0]
                # ys = agent_target_trajs.agents[0][:, ai, 1]
                if ((abs(agent_target_trajs.agents[0][0, ai, 3]) > 0.25) or 
                    (abs(agent_target_trajs.agents[0][0, ai, 4]) > 0.25)):
                  print("ai=", ai)
                  print(agent_target_trajs.agents[0][:, ai, :])

              print(">>> targets")
              print(targets.keys())
              print("ego_trajectory", ego_target_traj.data.size()) # [batch_siz, traj_size, 3]

              print(" ")

            # print("trajectory", ego_target_traj.data.size())

            cii += 1


    def run_dataloader_nuplan_ray(self) -> None:
        """
        Test dataloader using nuPlan DB.
        """
        scenario_builder = NuPlanScenarioBuilder(
            version=DATA_VERSION,
            data_root=os.getenv('NUPLAN_DATA_ROOT'))
        self.load_data(scenario_builder, RayDistributed())

    def run_dataloader_nuplan_sequential(self) -> None:
        """
        Test load_data using nuPlan DB using a sequential worker.
        """
        # I don't know why it can not works...
        scenario_builder = NuPlanScenarioBuilder(
            version=DATA_VERSION,
            data_root=os.getenv('NUPLAN_DATA_ROOT'))
        self.load_data(scenario_builder, Sequential())


def test_torch_device():
  # Device configuration
  device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
  print('set device=', device)

if __name__ == '__main__':
  test_torch_device()

  loader = DataLoader()
  loader.run_dataloader_nuplan_ray()
  # loader.run_dataloader_nuplan_sequential()
```

**Comment by patk-motional** [comment]
Hi, this is most likely due to agent padding. For example, let's say we want to build an agent feature with past agent history of 4 frames including the current time step, t_-3, t_-2, t_-1, t_0. If the agent at time t_0 for some reason does not exist at t_-1, the rest of the history is padded with the agent state at t_0. Hence, all the states from t_-3, ..., t_0 will be identical.

This is our current way to account for missing agents due to incomplete data or poor agent tracking. We're open to suggestions if you think there is a better way of doing this.

**Comment by ChenYingbing** [comment]
Hi, thank you for your patient response. It solves my confusion.

Regards,
Chen Yingbing.

**Comment by HaodongLi-Hub** [comment]
> Hi, thank you for your patient response. It solves my confusion.
> 
> Regards, Chen Yingbing.

Hi, may I ask where I can find the the nuplan dataset v0.2 or v0.2mini? I can not find it on the website, and it confuse me a lot. If you can share with me, I really apprieciate!
